File: python/09_prosody_similarity.py
"""
Prosody Similarity — Vocametrix API example

Compares pitch and intensity curves between a model (reference) speaker
and a learner speaker. Returns a similarity score and curve data.

Upload pattern: POST /api/assignFileId for both files
                → GET /api/calculate-prosody-similarity
Auth: X-API-Key header

Usage:
    python 09_prosody_similarity.py model.wav learner.wav
"""

import os
import argparse
import requests
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))
API_KEY = os.environ['VOCAMETRIX_API_KEY']
BASE_URL = 'https://platform.vocametrix.com'
HEADERS = {'X-API-Key': API_KEY}


def assign_file_id(audio_path: str) -> str:
    with open(audio_path, 'rb') as f:
        r = requests.post(
            f'{BASE_URL}/api/assignFileId',
            headers=HEADERS,
            files={'audio': f},
        )
    r.raise_for_status()
    return r.json()['fileId']


def calculate_prosody_similarity(model_path: str, learner_path: str) -> dict:
    print('Uploading model recording...')
    model_id = assign_file_id(model_path)
    print('Uploading learner recording...')
    learner_id = assign_file_id(learner_path)

    r = requests.get(
        f'{BASE_URL}/api/calculate-prosody-similarity',
        headers=HEADERS,
        params={'svFileId': model_id, 'csFileId': learner_id},
    )
    r.raise_for_status()
    return r.json()


def print_results(results: dict):
    print('\n=== Prosody Similarity Results ===')
    score = results.get('PROSODY_SIMILARITY_SCORE', results.get('similarity_score', 'N/A'))
    print(f"Overall similarity: {score}")

    for key in ('PITCH_CORRELATION', 'INTENSITY_CORRELATION', 'RHYTHM_SCORE'):
        if key in results:
            print(f"  {key}: {results[key]}")

    # Curve data is available for visualization
    pitch_model = results.get('PITCH_CURVE_MODEL', [])
    pitch_learner = results.get('PITCH_CURVE_LEARNER', [])
    if pitch_model and pitch_learner:
        print(f"\nPitch curve points: {len(pitch_model)} (model), {len(pitch_learner)} (learner)")
        print("  (Pass results to prosody_similarity_loop.py workflow for visualization)")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Prosody similarity analysis')
    parser.add_argument('model', help='Path to model/reference WAV')
    parser.add_argument('learner', help='Path to learner WAV')
    args = parser.parse_args()

    results = calculate_prosody_similarity(args.model, args.learner)
    print_results(results)

File: javascript/09-prosody-similarity.js
/**
 * Prosody Similarity — Vocametrix API example (Node.js)
 *
 * Compares pitch and intensity curves between a model (reference) speaker
 * and a learner speaker. Returns a similarity score and curve data.
 *
 * Upload pattern: POST /api/assignFileId for both files
 *                 → GET /api/calculate-prosody-similarity
 * Auth: X-API-Key header
 *
 * Usage:
 *   node 09-prosody-similarity.js model.wav learner.wav
 */

import fs from 'fs';
import FormData from 'form-data';
import 'dotenv/config';

const API_KEY = process.env.VOCAMETRIX_API_KEY;
const BASE_URL = 'https://platform.vocametrix.com';

async function assignFileId(audioPath, label) {
  console.log(`Uploading ${label}...`);
  const form = new FormData();
  form.append('audio', fs.createReadStream(audioPath));
  const res = await fetch(`${BASE_URL}/api/assignFileId`, {
    method: 'POST',
    headers: { 'X-API-Key': API_KEY, ...form.getHeaders() },
    body: form,
  });
  if (!res.ok) throw new Error(`assignFileId failed: ${await res.text()}`);
  return (await res.json()).fileId;
}

async function calculateProsodySimilarity(modelPath, learnerPath) {
  const svFileId = await assignFileId(modelPath, 'model');
  const csFileId = await assignFileId(learnerPath, 'learner');

  const res = await fetch(
    `${BASE_URL}/api/calculate-prosody-similarity?svFileId=${svFileId}&csFileId=${csFileId}`,
    { headers: { 'X-API-Key': API_KEY } },
  );
  if (!res.ok) throw new Error(`calculate-prosody-similarity failed: ${await res.text()}`);
  return res.json();
}

function printResults(results) {
  const score = results.PROSODY_SIMILARITY_SCORE ?? results.similarity_score ?? 'N/A';
  console.log('\n=== Prosody Similarity Results ===');
  console.log(`Overall similarity: ${score}`);
  for (const key of ['PITCH_CORRELATION', 'INTENSITY_CORRELATION', 'RHYTHM_SCORE']) {
    if (results[key] !== undefined) console.log(`  ${key}: ${results[key]}`);
  }
  const pitchModel = results.PITCH_CURVE_MODEL ?? [];
  const pitchLearner = results.PITCH_CURVE_LEARNER ?? [];
  if (pitchModel.length > 0) {
    console.log(`\nPitch curve points: ${pitchModel.length} (model), ${pitchLearner.length} (learner)`);
  }
}

const args = process.argv.slice(2);
if (args.length < 2) {
  console.error('Usage: node 09-prosody-similarity.js <model.wav> <learner.wav>');
  process.exit(1);
}
calculateProsodySimilarity(args[0], args[1]).then(printResults).catch(err => { console.error(err.message); process.exit(1); });
